src/Autofac/Core/Registration/DefaultRegisteredServicesTracker.cs

```diff
@@ -253,52 +233,80 @@ private ServiceRegistrationInfo GetInitializedServiceInfo(Service service)
         {
             var info = GetServiceInfo(service);
             if (info.IsInitialized)
-                return info;
-
-            if (!info.IsInitializing)
             {
-                BeginServiceInfoInitialization(service, info, _dynamicRegistrationSources);
+                return info;
             }
 
-            while (info.HasSourcesToQuery)
+            var succeeded = false;
+
+            Monitor.Enter(info);
```

Comment:
@alistairjevans  Hi and sorry for intruding,
The more robust `lock` pattern is using the overload with `Monitor.Enter(info, ref lockTaken)`, see details here http://www.albahari.com/threading/part2.aspx#_MonitorEnter_and_MonitorExit

Reply:
No worries, thanks for the tip! Change is in.